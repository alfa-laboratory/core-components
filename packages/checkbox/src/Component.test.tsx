import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Checkbox } from './index';

describe('Checkbox', () => {
    describe('Display tests', () => {
        it('should display correctly', () => {
            expect(render(<Checkbox label='label' hint='hint' />)).toMatchSnapshot();
        });

        it('should display indeterminate correctly', () => {
            expect(
                render(<Checkbox label='label' hint='hint' indeterminate={true} />),
            ).toMatchSnapshot();
        });
    });

    describe('Styles tests', () => {
        it('should set custom class', () => {
            const className = 'custom-class';
            const { container } = render(<Checkbox className={className} />);

            expect(container.firstElementChild).toHaveClass(className);
        });

        it('should set `checked` class', () => {
            const { container } = render(<Checkbox checked={true} />);

            expect(container.firstElementChild).toHaveClass('checked');
        });

        it('should set `disabled` class', () => {
            const { container } = render(<Checkbox disabled={true} />);

            expect(container.firstElementChild).toHaveClass('disabled');
        });

        it('should set `indeterminate` class', () => {
            const { container } = render(<Checkbox indeterminate={true} />);

            expect(container.firstElementChild).toHaveClass('indeterminate');
        });
    });

    describe('Attributes tests', () => {
        it('should set `data-test-id` atribute', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Checkbox dataTestId={dataTestId} />);

            expect(getByTestId(dataTestId).tagName).toBe('INPUT');
        });

        it('should set disabled attribute', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Checkbox disabled={true} dataTestId={dataTestId} />);

            expect(getByTestId(dataTestId)).toBeDisabled();
        });

        it('should set checked attribute', () => {
            const dataTestId = 'test-id';
            const { getByTestId } = render(<Checkbox checked={true} dataTestId={dataTestId} />);

            expect(getByTestId(dataTestId)).toBeChecked();
        });
    });

    describe('Render tests', () => {
        test('should unmount without errors', () => {
            const { unmount } = render(<Checkbox />);

            expect(unmount).not.toThrowError();
        });
    });

    describe('Interaction tests', () => {
        test('should call `onChange` prop if not disabled', () => {
            const cb = jest.fn();

            const { container } = render(<Checkbox onChange={cb} checked={true} />);

            if (container.firstElementChild) {
                fireEvent.click(container.firstElementChild);
            }

            expect(cb).toBeCalledTimes(1);
        });

        test('should not call `onChange` prop if disabled', () => {
            const cb = jest.fn();

            const { container } = render(<Checkbox onChange={cb} disabled={true} />);

            if (container.firstElementChild) {
                fireEvent.click(container.firstElementChild);
            }

            expect(cb).not.toBeCalled();
        });

        test('should not call `onChange` prop if disabled and checked', () => {
            const cb = jest.fn();

            const { container } = render(<Checkbox onChange={cb} checked={true} disabled={true} />);

            if (container.firstElementChild) {
                fireEvent.click(container.firstElementChild);
            }

            expect(cb).not.toBeCalled();
        });
    });
});
