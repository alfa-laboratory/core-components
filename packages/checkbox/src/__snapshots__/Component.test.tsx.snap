// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Display tests should display correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="component"
      >
        <span
          class="box"
        >
          <input
            type="checkbox"
          />
        </span>
        <span
          class="content"
        >
          <span
            class="label"
          >
            label
          </span>
          <span
            class="hint"
          >
            hint
          </span>
        </span>
      </label>
    </div>
  </body>,
  "container": <div>
    <label
      class="component"
    >
      <span
        class="box"
      >
        <input
          type="checkbox"
        />
      </span>
      <span
        class="content"
      >
        <span
          class="label"
        >
          label
        </span>
        <span
          class="hint"
        >
          hint
        </span>
      </span>
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Checkbox Display tests should display indeterminate correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <label
        class="component indeterminate"
      >
        <span
          class="box"
        >
          <input
            type="checkbox"
          />
          <svg
            class="indeterminateIcon"
            focusable="false"
            viewBox="0 0 18 18"
          >
            <path
              d="M4 8h10v2H4z"
              fill="#fff"
            />
          </svg>
        </span>
        <span
          class="content"
        >
          <span
            class="label"
          >
            label
          </span>
          <span
            class="hint"
          >
            hint
          </span>
        </span>
      </label>
    </div>
  </body>,
  "container": <div>
    <label
      class="component indeterminate"
    >
      <span
        class="box"
      >
        <input
          type="checkbox"
        />
        <svg
          class="indeterminateIcon"
          focusable="false"
          viewBox="0 0 18 18"
        >
          <path
            d="M4 8h10v2H4z"
            fill="#fff"
          />
        </svg>
      </span>
      <span
        class="content"
      >
        <span
          class="label"
        >
          label
        </span>
        <span
          class="hint"
        >
          hint
        </span>
      </span>
    </label>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
